var/spack/repos/builtin/packages/trilinos/package.py

```diff
@@ -270,15 +265,23 @@ class Trilinos(CMakePackage, CudaPackage):
         conflicts('+teko')
         conflicts('+zoltan2')
 
+    with when('+teko'):
+        conflicts('~amesos')
+        conflicts('~anasazi')
+        conflicts('~aztec')
+        conflicts('~ifpack')
+        conflicts('~ml')
+        conflicts('~stratimikos')
+        conflicts('@:12 gotype=long')
+
     conflicts('+basker', when='~amesos2')
     conflicts('+ifpack2', when='~belos')
     conflicts('+intrepid', when='~sacado')
     conflicts('+intrepid', when='~shards')
     conflicts('+intrepid2', when='~shards')
     conflicts('+isorropia', when='~zoltan')
+    conflicts('+muelu', when='~tpetra')
```

Comment:
Based on the discussion, this should also have a version restriction for greater than 13.0.1 right?

Reply:
@sethrj I'm not sure how long MueLu has failed to build without Tpetra + Kokkos. I don't think that particular combination has been under test (MueLu Epetra only). In the other PR about it failing, they didn't want to build MueLu, it was just set to ON by default which is why they ran into that problem. I'm not sure what version of Trilinos we would first see this fail to build. I can check.

Reply:
I see, I thought that it had worked previously based on this conversation. Prior to that I had always understood muelu to be a second-generation package, but with all the adapter layers to adapters in trilinos I can't be sure. If someone thinks an older verion builds successfully we can always add support back in. Thanks!

Reply:
Just to follow up here.

Prior to and including 12.12.1, running `spack -v -k install trilinos~kokkos~tpetra~ifpack2~amesos2~belos` results in:
```
  >> 2269    /var/folders/zk/1zr9bdg53s3_d7qxmv2m5f80002l2h/T/pakuber/spack-stage/spack-stage-trilinos-12.12.1-4xpgfmh4khm6zb3gfo2bwxhojg
             k52mbr/spack-src/packages/ml/src/Coarsen/ml_aggregate.c:2097:3: error: implicit declaration of function 'ML_Operator_Copy_St
             atistics' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
     2270      ML_Operator_Copy_Statistics(mat,permuted_Amat);
     2271      ^
     2272    /var/folders/zk/1zr9bdg53s3_d7qxmv2m5f80002l2h/T/pakuber/spack-stage/spack-stage-trilinos-12.12.1-4xpgfmh4khm6zb3gfo2bwxhojg
             k52mbr/spack-src/packages/ml/src/Coarsen/ml_aggregate.c:2097:3: note: did you mean 'ML_Operator_ReportStatistics'?
     2273    /var/folders/zk/1zr9bdg53s3_d7qxmv2m5f80002l2h/T/pakuber/spack-stage/spack-stage-trilinos-12.12.1-4xpgfmh4khm6zb3gfo2bwxhojg
             k52mbr/spack-src/packages/ml/src/Operator/ml_op_utils.h:110:13: note: 'ML_Operator_ReportStatistics' declared here
     2274    extern void ML_Operator_ReportStatistics(ML_Operator *mat, char *appendlabel,
     2275                ^
  >> 2276    /var/folders/zk/1zr9bdg53s3_d7qxmv2m5f80002l2h/T/pakuber/spack-stage/spack-stage-trilinos-12.12.1-4xpgfmh4khm6zb3gfo2bwxhojg
             k52mbr/spack-src/packages/ml/src/Coarsen/ml_aggregate.c:2416:5: error: implicit declaration of function 'ML_Operator_Copy_St
             atistics' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
     2277        ML_Operator_Copy_Statistics(Pmat,newP);
     2278        ^
     2279    2 errors generated.
  >> 2280    make[2]: *** [packages/ml/src/CMakeFiles/ml.dir/Coarsen/ml_aggregate.c.o] Error 1
     2281    make[2]: *** Waiting for unfinished jobs....
     2282    /var/folders/zk/1zr9bdg53s3_d7qxmv2m5f80002l2h/T/pakuber/spack-stage/spack-stage-trilinos-12.12.1-4xpgfmh4khm6zb3gfo2bwxhojg
             k52mbr/spack-src/packages/ml/src/Coarsen/ml_agg_min_energy.cpp:48:10: warning: 'register' storage class specifier is depreca
             ted and incompatible with C++17 [-Wdeprecated-register]
     2283      double register dtemp;
     2284             ^~~~~~~~~
     2285    1 warning generated.
  >> 2286    make[1]: *** [packages/ml/src/CMakeFiles/ml.dir/all] Error 2
  >> 2287    make: *** [all] Error 2
```
which is an ml error (not MueLu). I then also use `~ml`, and then it builds successfully.

12.14.1 builds successfully.

12.18.1 fails with:
```
  >> 2108    /var/folders/zk/1zr9bdg53s3_d7qxmv2m5f80002l2h/T/pakuber/spack-stage/spack-stage-trilinos-12.18.1-6elwskl2ck6yc2wyuz2elrz62t
             zmwhzt/spack-src/packages/xpetra/src/Headers/Xpetra_TpetraConfigDefs.hpp:52:2: error: This file should be included only if H
             AVE_XPETRA_TPETRA is defined.
     2109    #error This file should be included only if HAVE_XPETRA_TPETRA is defined.
     2110     ^
     2111    In file included from /var/folders/zk/1zr9bdg53s3_d7qxmv2m5f80002l2h/T/pakuber/spack-stage/spack-stage-trilinos-12.18.1-6elw
             skl2ck6yc2wyuz2elrz62tzmwhzt/spack-src/packages/xpetra/src/Map/Xpetra_EpetraMapFactory.cpp:50:
     2112    In file included from /private/var/folders/zk/1zr9bdg53s3_d7qxmv2m5f80002l2h/T/pakuber/spack-stage/spack-stage-trilinos-12.1
             8.1-6elwskl2ck6yc2wyuz2elrz62tzmwhzt/spack-build-6elwskl/packages/xpetra/src/Xpetra_TpetraMap.hpp:1:
  >> 2113    /var/folders/zk/1zr9bdg53s3_d7qxmv2m5f80002l2h/T/pakuber/spack-stage/spack-stage-trilinos-12.18.1-6elwskl2ck6yc2wyuz2elrz62t
             zmwhzt/spack-src/packages/xpetra/src/Map/Xpetra_TpetraMap_decl.hpp:53:10: fatal error: 'Tpetra_Map.hpp' file not found
     2114    #include <Tpetra_Map.hpp>
     2115             ^~~~~~~~~~~~~~~~
     2116    2 errors generated.
  >> 2117    make[2]: *** [packages/xpetra/src/CMakeFiles/xpetra.dir/Map/Xpetra_EpetraMapFactory.cpp.o] Error 1
     2118    make[2]: *** Waiting for unfinished jobs....
     2119    /var/folders/zk/1zr9bdg53s3_d7qxmv2m5f80002l2h/T/pakuber/spack-stage/spack-stage-trilinos-12.18.1-6elwskl2ck6yc2wyuz2elrz62t
             zmwhzt/spack-src/packages/xpetra/src/CrsGraph/Xpetra_EpetraCrsGraph.cpp:268:16: warning: explicit instantiation of 'EpetraCr
             sGraphT<int, Kokkos::Compat::KokkosSerialWrapperNode>' that occurs after an explicit specialization has no effect [-Winstant
             iation-after-specialization]
     2120    template class EpetraCrsGraphT<int, default_node_type >;
```
which is an Xpetra issue, and then 13.0.0 and above fails because of missing Kokkos.

To summarize, MueLu requires Xpetra and this will fail without Tpetra for Trilinos 12.18.1, and MueLu will fail without Kokkos for 13.0.0 and above.

It would have been possible to ease the restrictions for `~tpetra`  to conflict with only for Trilinos 12.18.1 and beyond.